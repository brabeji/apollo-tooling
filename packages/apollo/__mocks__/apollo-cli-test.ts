import * as fs from "fs";
import Nock from "@fancy-test/nock";
import * as Test from "@oclif/test";
export { expect } from "@oclif/test";
import { mockConsole } from "heroku-cli-util";
// import { userInfo } from "os";
// import { vol } from "memfs";
// import { ufs } from "unionfs";

const time = label => {
  let start = +new Date();

  return {
    async run() {},
    finally() {
      console.log(`${label || "TIME"}:`, +new Date() - start, "ms");
    }
  };
};

const debug = fn => {
  return {
    async run() {
      fn();
    },
    finally() {}
  };
};

const setupFS = (files: Record<string, string>) => {
  let dir;
  return {
    async run(ctx: any, ...rest) {
      // make a random remp dir & chdir into it
      dir = fs.mkdtempSync("__tmp__");
      process.chdir(dir);
      // fill the dir with `files`
      Object.keys(files).forEach(key => {
        fs.writeFileSync(key, files[key]);
      });
    },
    finally(ctx: any) {
      const postFiles = fs.readdirSync("./");
      // delele all files
      postFiles.forEach(key => {
        fs.unlinkSync(key);
      });
      // go up one level & delete the temp dir
      process.chdir("../");
      fs.rmdirSync(dir);
    }
  };
};

export const test = Test.test
  .register("nock", Nock)
  .register("fs", setupFS)
  .register("timing", time)
  .register("debug", debug);

mockConsole();
