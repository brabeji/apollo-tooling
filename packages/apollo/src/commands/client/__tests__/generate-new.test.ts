// jest.mock("apollo-codegen-core/lib/localfs", () => {
//   return require("../../../../__mocks__/mockFs");
// });

// this is because of herkou-cli-utils hacky mocking system on their console logger
import { stdout, mockConsole } from "heroku-cli-util";
import * as path from "path";
import * as fs from "fs";
import { test as setup } from "apollo-cli-test";
import {
  introspectionQuery,
  print,
  execute,
  buildSchema,
  graphql
} from "graphql";
import gql from "graphql-tag";

const test = setup.do(() => mockConsole());

// helper function to resolve files from the actual filesystem
const resolveFiles = opts => {
  let files = {};
  Object.keys(opts).map(key => {
    files[key] = fs.readFileSync(path.resolve(__dirname, opts[key]), {
      encoding: "utf-8"
    });
  });

  return files;
};

const {
  graphQLSchema,
  simpleQuery,
  otherQuery,
  clientSideSchema,
  clientSideSchemaQuery,
  clientSideOnlySchema,
  clientSideOnlyQuery
} = resolveFiles({
  graphQLSchema: "../../service/__tests__/fixtures/schema.graphql",
  simpleQuery: "./fixtures/simpleQuery.graphql",
  otherQuery: "./fixtures/otherQuery.graphql",
  clientSideSchema: "./fixtures/clientSideSchema.graphql",
  clientSideSchemaQuery: "./fixtures/clientSideSchemaQuery.graphql",
  clientSideOnlySchema: "./fixtures/clientSideOnlySchema.graphql",
  clientSideOnlyQuery: "./fixtures/clientSideOnlyQuery.graphql"
});

const fullSchema = execute(buildSchema(graphQLSchema), gql(introspectionQuery))
  .data;

const clientSideSchemaTag = `
  gql\`
  ${clientSideSchema}
  \`
`;

const serverSideSchemaTag = `
  gql\`
  type Query {
    hello: String!
    serverSideField: ServerField!
  }

  type ServerField {
    serverData: String!
  }

  type RemovedField {
    id: ID!
    name: RemovedType
  }

  type RemovedType {
    fieldName: String
  }
  \`
  `;

const apolloConfig = `
  module.exports = {
    client: {
      includes: ["./**.graphql"],
      service: {
        name: "my-service-name",
        localSchemaFile: "./schema.json"
      }
    }
  }
`;

describe("client:codegen", () => {
  test
    .stdout({ print: true })
    .fs({
      "./schema.json": JSON.stringify(fullSchema.__schema),
      "./queryOne.graphql": simpleQuery.toString(),
      "./apollo.config.js": `
        module.exports = {
          client: {
            includes: ["./**.graphql"],
            service: {
              name: "my-service-name",
              localSchemaFile: "./schema.json"
            }
          }
        }
      `
    })
    .command([
      "client:codegen",
      "API.swift",
      "--config=apollo.config.js",
      "--target=swift"
    ])
    .it("writes swift types from local schema", () => {
      expect(fs.readFileSync("API.swift").toString()).toMatchSnapshot();
    });

  test
    .only()
    .fs({
      "schema.graphql": graphQLSchema,
      "queryOne.graphql": simpleQuery.toString(),
      "apollo.config.js": `
        module.exports = {
          client: {
            includes: ["./**.graphql"],
            service: {
              name: "my-service-name",
              localSchemaFile: "./schema.graphql"
            }
          }
        }
      `
    })
    .command([
      "client:codegen",
      "API.swift",
      "--config=apollo.config.js",
      "--target=swift"
    ])
    .it("writes swift types from local schema in a graphql file", () => {
      expect(fs.readFileSync("API.swift").toString()).toMatchSnapshot();
    });
});
