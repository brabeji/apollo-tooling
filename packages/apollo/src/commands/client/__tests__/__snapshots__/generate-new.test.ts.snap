// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client:codegen writes swift types from local schema 1`] = `
"//  This file was automatically generated and should not be edited.

import Apollo

public final class SimpleQueryQuery: GraphQLQuery {
  public let operationDefinition =
    \\"query SimpleQuery {\\\\n  hello\\\\n}\\"

  public init() {
  }

  public struct Data: GraphQLSelectionSet {
    public static let possibleTypes = [\\"Query\\"]

    public static let selections: [GraphQLSelection] = [
      GraphQLField(\\"hello\\", type: .nonNull(.scalar(String.self))),
    ]

    public private(set) var resultMap: ResultMap

    public init(unsafeResultMap: ResultMap) {
      self.resultMap = unsafeResultMap
    }

    public init(hello: String) {
      self.init(unsafeResultMap: [\\"__typename\\": \\"Query\\", \\"hello\\": hello])
    }

    public var hello: String {
      get {
        return resultMap[\\"hello\\"]! as! String
      }
      set {
        resultMap.updateValue(newValue, forKey: \\"hello\\")
      }
    }
  }
}"
`;
